File: web/src/api/channel.ts
import axios from 'axios';

export interface ChannelData {
  id: string;
  workspace_id: string;
  name: string;
  description: string;
  created_at: string;
  created_by: string;
  updated_at: string;
}

export const getChannels = async (workspaceId: string) => {
  const response = await axios.get(`/api/v1/workspaces/${workspaceId}/channels`, { withCredentials: true });
  return response.data as ChannelData[];
};

export const createChannel = async (workspaceId: string, data: { name: string; description?: string }) => {
  const response = await axios.post(`/api/v1/workspaces/${workspaceId}/channels`, {
    name: data.name,
    description: data.description ?? '',
  });
  return response.data as ChannelData;
};

export const updateChannel = async (workspaceId: string, channelId: string, data: { name: string; description?: string }) => {
  const response = await axios.put(`/api/v1/workspaces/${workspaceId}/channels/${channelId}`, {
    name: data.name,
    description: data.description ?? '',
  });
  return response.data as ChannelData;
};

export const deleteChannel = async (workspaceId: string, channelId: string) => {
  const response = await axios.delete(`/api/v1/workspaces/${workspaceId}/channels/${channelId}`);
  return response.data;
};

File: api/internal/db/sqlitedb/channel.go
package sqlitedb

import (
	"context"

	"github.com/notomate/notomate/internal/model"
	"gorm.io/gorm"
)

func (s SqliteDB) CreateChannel(c model.Channel) error {
	return gorm.G[model.Channel](s.getDB()).Create(context.Background(), &c)
}

func (s SqliteDB) UpdateChannel(c model.Channel) error {
	_, err := gorm.G[model.Channel](s.getDB()).
		Where("id = ?", c.ID).
		Select("name", "description", "updated_at", "updated_by").
		Updates(context.Background(), c)
	return err
}

func (s SqliteDB) DeleteChannel(c model.Channel) error {
	_, err := gorm.G[model.Channel](s.getDB()).Where("id = ?", c.ID).Delete(context.Background())
	return err
}

func (s SqliteDB) FindChannel(c model.Channel) (model.Channel, error) {
	channel, err := gorm.
		G[model.Channel](s.getDB()).
		Where("id = ?", c.ID).
		Take(context.Background())

	return channel, err
}

func (s SqliteDB) FindChannels(f model.ChannelFilter) ([]model.Channel, error) {
	var channels []model.Channel

	query := s.getDB().Model(&model.Channel{})

	if f.WorkspaceID != "" {
		query = query.Where("workspace_id = ?", f.WorkspaceID)
	}

	err := query.Order("created_at ASC").Find(&channels).Error

	return channels, err
}
